Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,393 @@
+    }
+
+    override func layoutAttributesForElementsInRect(rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
+        var allAttributes = [UICollectionViewLayoutAttributes]()
+        for i in 0 ..< cellCount {
+            let indexPath = NSIndexPath(forRow: i, inSection: 0)
+            let attr = self.computeLayoutAttributesForCellAtIndexPath(indexPath)
+            allAttributes.append(attr)
+        }
+        return allAttributes
+    }
+
+    override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
+        return self.computeLayoutAttributesForCellAtIndexPath(indexPath)
+    }
+
+    override func shouldInvalidateLayoutForBoundsChange(newBounds: CGRect) -> Bool {
+        if newBounds.width < 320 {
+            return false
+        } else {
+            return true
+        }
+    }
+
+    func computeLayoutAttributesForCellAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes {
+        let row = indexPath.row
+        let bounds = self.collectionView!.bounds
+
+        let verticalItemsCount = maxVerticalItemsCount()
+        let horizontalItemsCount = maxHorizontalItemsCount()
+
+
+        let itemsPerPage = verticalItemsCount * horizontalItemsCount
+
+        let columnPosition = row % horizontalItemsCount
+        let rowPosition = (row / horizontalItemsCount) % verticalItemsCount
+        let itemPage = Int(floor(Double(row)/Double(itemsPerPage)))
+
+        let attr = UICollectionViewLayoutAttributes(forCellWithIndexPath: indexPath)
+
+        var frame = CGRectZero
+        frame.origin.x = CGFloat(itemPage) * bounds.size.width + CGFloat(columnPosition) * (itemSize.width + insets.left) + insets.left
+        frame.origin.y = CGFloat(rowPosition) * (itemSize.height + insets.top) + insets.top
+        frame.size = itemSize
+        attr.frame = frame
+        
+        return attr
+    }
+}
+
+/*
+    Defines the number of items to show in topsites for different size classes.
+*/
+struct ASTopSiteSourceUX {
+    static let verticalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 1, UIUserInterfaceSizeClass.Regular : 2]
+    static let horizontalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 3, UIUserInterfaceSizeClass.Regular : 5]
+    static let maxNumberOfPages = 2
+    static let CellIdentifier = "TopSiteCell"
+}
+
+protocol ASHorizontalScrollDelegate {
+    func numberOfVerticalItems(estimatedItems: Int) -> Int
+    func numberOfHorizontalItems(estimatedItems: Int) -> Int
+}
+
+/*
+ This Delegate/DataSource is used to manage the ASHorizontalScrollCell's UICollectionView. 
+ This is left generic enough for it to be re used for other parts of Activity Stream.
+ */
+
+class ASHorizontalScrollSource: NSObject, UICollectionViewDelegate, UICollectionViewDataSource, ASHorizontalScrollDelegate {
+
+    var content: [TopSiteItem] = []
+    var urlPressedHandler: ((NSURL) -> Void)?
+    var pageChangedHandler: ((CGFloat) -> Void)?
```

Comment:
Sorry I think I was getting confused about how these handlers and delegates work. I took a closer look and I think I understand whats going on here. The handlers should be fine but I found a couple of things that were a bit confusing when reading through this.

1. It might be more clear to also define a `ASHorizontalScrollDataSource` along with the `ASHorizontalScrollDelegate`. Since you're following the UICollectionViewDS/Del convention, smooshing both of these under a 'delegate' is a bit of a misnomer.

2. I would also clarify what the delegate/data source do by separating them out into their own extensions off of `ASHorizontalScrollSource`. It might be a bit overkill considering that the `ASHorizontalScrollView` at the moment requires a `UICollectionView` so you can probably have the protocols extend from the `UICollectionView` ones.

3. Minor thing but I would rename `topSiteHandler` to something more reflective of how it's both a datasource and delegate for the top sites cell since you use the handler prefix for callback methods inside the `topSiteHandler`. Maybe `topSiteSource`?

I also tried to think of how the handlers would work as delegate methods but it's a bit strange because the `ActivityStreamPanel` needs the tap event while the cell itself needs the page change event. One option could be to define different delegates `ASHorizontalScrollInteractionDelegate` / `ASHorizontalScrollPageDelegate` but considering that we have one method for each one that's a bit crazy and pointless so stick with the handlers 👍